inbm/dispatcher-agent/dispatcher/dispatcher_class.py

```diff
@@ -186,7 +190,13 @@ def _sig_handler(signo, frame) -> None:
                 if active_count() - active_start_count < self._thread_count:
                     worker = Thread(target=handle_updates, args=(self,))
                     worker.setDaemon(True)
+                    self._thread_list.append(worker)
                     worker.start()
+
+            # Periodically check if processes have finished. If process finished, remove it from the list.
+            for thread in self._thread_list:
+                if not thread.is_alive():
+                    self._thread_list.remove(thread)
```

Comment:
modifying a list while iterating over it can lead to unexpected behaviors (can skip elements or have runtime errors). instead, iterate over a shallow copy of the list --
```
for thread in self._thread_list[:]:
    if not thread.is_alive():
        self._thread_list.remove(thread)
```